include/pmacc/particles/algorithm/ForEach.hpp

```diff
@@ -0,0 +1,194 @@
+ *
+ * PMacc is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License and the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * and the GNU Lesser General Public License along with PMacc.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "pmacc/Environment.hpp"
+#include "pmacc/particles/frame_types.hpp"
+#include "pmacc/mappings/kernel/AreaMapping.hpp"
+#include "pmacc/mappings/threads/ForEachIdx.hpp"
+#include "pmacc/mappings/threads/IdxConfig.hpp"
+#include "pmacc/mappings/threads/WorkerCfg.hpp"
+
+
+namespace pmacc
+{
+namespace particles
+{
+namespace algorithm
+{
+namespace acc
+{
+namespace detail
+{
+
+    /** operate on particles of a species
+     *
+     * @tparam T_numWorkers number of workers
+     */
+    template< uint32_t T_numWorkers >
+    struct ForEachParticle
+    {
+        /** operate on particles
+         *
+         * @tparam T_Acc alpaka accelerator type
+         * @tparam T_Functor type of the functor to operate on a particle
+         * @tparam T_Mapping mapping functor type
+         * @tparam T_ParBox pmacc::ParticlesBox, type of the species box
+         *
+         * @param acc alpaka accelerator
+         * @param functor functor to operate on a particle
+         *                must fulfill the interface pmacc::functor::Interface<F, 1u, void>
+         * @param mapper functor to map a block to a supercell
+         * @param pb particles species box
+         */
+        template<
+            typename T_Acc,
+            typename T_Functor,
+            typename T_Mapping,
+            typename T_ParBox
+        >
+        DINLINE void operator()(
+            T_Acc const & acc,
+            T_Functor functor,
+            T_Mapping const mapper,
+            T_ParBox pb
+        ) const
+        {
+            using namespace mappings::threads;
+
+            using SuperCellSize = typename T_ParBox::FrameType::SuperCellSize;
+            constexpr uint32_t dim = SuperCellSize::dim;
+            constexpr uint32_t frameSize = pmacc::math::CT::volume< SuperCellSize >::type::value;
+            constexpr uint32_t numWorkers = T_numWorkers;
+
+            uint32_t const workerIdx = threadIdx.x;
+
+            DataSpace< dim > const superCellIdx(
+                mapper.getSuperCellIndex( DataSpace< dim >( blockIdx ) )
+            );
+
+            auto const & superCell = pb.getSuperCell( superCellIdx );
+            uint32_t const numPartcilesInSupercell = superCell.getNumParticles();
+
+
+            // end kernel if we have no particles
+            if( numPartcilesInSupercell == 0 )
+                return;
+
+            using FramePtr = typename T_ParBox::FramePtr;
+            FramePtr frame = pb.getLastFrame( superCellIdx );
```

Comment:
Bug: to match the condition in the loop, should here start from the first frame and inside the loop go to the next frame.